trin-core/src/jsonrpc/types.rs

```diff
@@ -239,12 +239,66 @@ impl TryFrom<[&Value; 2]> for FindContentParams {
     }
 }
 
-pub struct OfferParams {
+pub struct OfferParams<TContentKey> {
+    pub content_key: TContentKey,
+    pub content: Vec<u8>,
+}
+
+impl<TContentKey: OverlayContentKey> TryFrom<Params> for OfferParams<TContentKey> {
+    type Error = ValidationError;
+
+    fn try_from(params: Params) -> Result<Self, Self::Error> {
+        match params {
+            Params::Array(val) => match val.len() {
+                2 => OfferParams::<TContentKey>::try_from([&val[0], &val[1]]),
+                _ => Err(ValidationError::new("Expected 2 params")),
+            },
+            _ => Err(ValidationError::new("Expected array of params")),
+        }
+    }
+}
+
+impl<TContentKey: OverlayContentKey> TryFrom<[&Value; 2]> for OfferParams<TContentKey> {
+    type Error = ValidationError;
+
+    fn try_from(params: [&Value; 2]) -> Result<Self, Self::Error> {
+        let content_key = params[0]
+            .as_str()
+            .ok_or_else(|| ValidationError::new("Empty content key param"))?;
+        let content_key = match hex_decode(content_key) {
+            Ok(val) => match TContentKey::try_from(val) {
+                Ok(val) => val,
+                Err(_) => return Err(ValidationError::new("Unable to decode content_key")),
+            },
+            Err(err) => {
+                let mut ve = ValidationError::new("Cannot convert content_key hex to bytes");
+                ve.add_param(
+                    std::borrow::Cow::Borrowed("hex_decode_exception"),
+                    &err.to_string(),
+                );
+                return Err(ve);
+            }
+        };
+        let content = params[1]
+            .as_str()
+            .ok_or_else(|| ValidationError::new("Empty content param"))?;
+        let content = match hex_decode(content) {
+            Ok(val) => val,
+            Err(_) => return Err(ValidationError::new("Unable to decode content")),
+        };
+        Ok(Self {
+            content_key,
+            content,
+        })
```

Comment:
There's a whole lot of copy/paste in here from `StoreParams`. Would probably be nice to reduce that